spark/src/main/scala/org/apache/spark/sql/delta/actions/TableFeatureSupport.scala

```diff
@@ -368,7 +388,19 @@ trait TableFeatureSupport { this: Protocol =>
    * features. After we remove the last native feature we downgrade the protocol to (1, 1).
    */
   def downgradeProtocolVersionsIfNeeded: Protocol = {
-    if (!readerAndWriterFeatures.forall(_.isLegacyFeature)) return this
+    if (nativeReaderAndWriterFeatures.nonEmpty) {
+      val (minReaderVersion, minWriterVersion) =
+        TableFeatureProtocolUtils.minimumRequiredVersions(nativeReaderAndWriterFeatures)
+      // It is guaranteed by the definitions of WriterFeature and ReaderFeature, that we cannot
+      // end up with invalid protocol versions such as (3, 3). Nevertheless,
+      // we double check it here.
+      val newProtocol =
+      Protocol(minReaderVersion, minWriterVersion).withFeatures(readerAndWriterFeatures)
+      require(
```

Comment:
`assert`? This method takes no arguments, so you can hardly require something of them.